baking/src/tezos_baking/provider.py

```diff
@@ -0,0 +1,250 @@
+            # it could happen that `snapshot_version` field is not supplied by provider
+            # e.g. marigold snapshots don't supply it
+            lambda major, minor, rc, snapshot_version: snapshot_version
+            and compatible_snapshot_version - snapshot_version <= 2
+        )
+
+        non_rc_on_stable_pred = lambda major, minor, rc, snapshot_version: (
+            rc_version is None and rc is None
+        ) or (rc_version is not None)
+
+        preds = [
+            exact_version_pred,
+            compose_pred(
+                non_rc_on_stable_pred,
+                compatible_version_pred,
+                sum_pred(
+                    compose_pred(
+                        exact_major_version_pred,
+                        exact_minor_version_pred,
+                        less_rc_version_pred,
+                    ),
+                    compose_pred(
+                        exact_major_version_pred,
+                        less_minor_version_pred,
+                        non_rc_version_pred,
+                    ),
+                ),
+            ),
+            compose_pred(
+                non_rc_on_stable_pred,
+                compatible_version_pred,
+            ),
+        ]
+
+        return next(
+            (
+                snapshot
+                for snapshot in map(
+                    lambda pred: find_snapshot(pred),
+                    preds,
+                )
+                if snapshot is not None
+            ),
+            None,
+        )
+
+    def get_snapshot_metadata(self, network, history_mode):
+        snapshot_array = None
+        with urllib.request.urlopen(self.metadata_url) as url:
+            snapshot_array = json.load(url)["data"]
+        snapshot_array.sort(reverse=True, key=lambda x: x["block_height"])
+        return self.extract_relevant_snapshot(snapshot_array, network, history_mode)
+
+
+class TzInit(Provider):
+    def get_filesize(self, url):
+        request = urllib.request.Request(
+            url, headers=http_request_headers, method="HEAD"
+        )
+        content_length = next(
+            (
+                header[1]
+                for header in urllib.request.urlopen(request).info()._headers
+                if header[0] == "Content-Length"
+            ),
+            None,
+        )
+        if content_length is not None:
+            suffixes = ["B", "KB", "MB", "GB", "TB", "PB"]
+            nbytes = int(content_length)
+            i = 0
+            while nbytes >= 1024 and i < len(suffixes) - 1:
+                nbytes /= 1024.0
+                i += 1
+            f = ("%.2f" % nbytes).rstrip("0").rstrip(".")
+            return "%s%s" % (f, suffixes[i])
+        return content_length
+
+    def get_snapshot_metadata(self, network, history_mode):
+        history_mode = "full" if history_mode == "archive" else history_mode
+        self.metadata_url = (
+            f"https://snapshots.eu.tzinit.org/{network}/{history_mode}.json"
```

Comment:
Is it possible to add one more step that asks user to select a closest region? It would be nice to utilize this feature of `tzinit` provider